tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,252 @@
+            Err(err) => match err {
+                // Retry if it's a transport failure, might be due to intermittent connection issues
+                tough::error::Error::Transport { .. } => {
+                    error!(
+                        "Repo target fetch failure on attempt {} out of {}",
+                        attempt, max_attempts
+                    );
+                    if attempt >= max_attempts {
+                        return Err(repo_error::Error::ReadTarget {
+                            target,
+                            source: err,
+                        });
+                    }
+                }
+                // For other errors we should just return immediately and not retry
+                err => {
+                    return Err(repo_error::Error::ReadTarget {
+                        target,
+                        source: err,
+                    })
+                }
+            },
+            Ok(target_reader) => return Ok(target_reader),
+        };
+    }
+
+    Ok(None)
+}
+
+async fn attempt_download_target(
+    mut reader: impl Read + Send,
+    target: String,
+) -> Result<(), Error> {
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        info!("Downloading target: {}", target);
+        // tough's `Read` implementation validates the target as it's being downloaded
+        if let Err(err) = io::copy(&mut reader, &mut io::sink()) {
+            error!(
+                "Download error on attempt {} out of {}: {}",
+                attempt, max_attempts, err
+            );
+            if attempt >= max_attempts {
+                return Err(error::Error::TargetDownload {
+                    target,
+                    source: err,
+                });
+            }
+            continue;
+        }
+        // Successfully downloaded target
+        break;
+    }
+
+    Ok(())
+}
+
+/// Randomly samples specified percentage of listed targets in the TUF repo and tries to retrieve them.
+async fn retrieve_percentage_of_targets<T: 'static>(
+    repo: &Repository<'_, T>,
+    percentage: u8,
+) -> Result<(), Error>
+where
+    T: tough::Transport,
+    <T as tough::Transport>::Stream: std::marker::Send,
+{
+    let targets = &repo.targets().signed.targets;
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+    let targets = targets.keys().cloned().collect::<Vec<String>>();
+
+    let mut tasks = Vec::new();
+    for target in targets
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+    {
+        let reader =
+            attempt_read_target(&repo, target.to_string())?.context(error::TargetMissing {
+                target: target.clone(),
+            })?;
+
+        tasks.push(tokio::spawn(async move {
+            attempt_download_target(reader, target.to_string()).await
+        }));
```

Comment:
I think we may want to do both `attempt_read_target` and `attempt_download_target` in the spawned task, so we can add delays between retry attempts in both; it's not as effective to try immediately, we could wait a second.

(I wonder if the functions should be combined, actually.  If a download fails, reusing the same reader object seems dangerous, right?  `Read`ers don't seek back to the start, so it'd only work if the client handled range requests, I think?  Not all `Read`ers would still be valid after a failure.)

Reply:
I think tough's `HttpTransport` handles retries and sends range request as a consequence of @webern 's work in https://github.com/awslabs/tough/pull/143.

So actually I'm not sure if it's worth implementing the retries here at all...

Reply:
I'd be OK with removing it; as long as it's going through the transport, that's a better spot for the logic.  If you'd prefer not to, I think it'd make sense to combine read/download, so we don't reuse readers.